### ISSUE TITLE
Add link to Odesly itself

### ISSUE THREAD

**Issue by AlexWayfer** [9dogs/tg-odesli-bot#7]
Hello!

For now result of bot looks like:

> Animadrop, Ben Walter, Aeris - Thin Air
> [Deezer](https://www.deezer.com/track/738571332) | [Spotify](https://open.spotify.com/track/4HhvD1HVtvZ0eaCP0Sp0F0) | [Apple Music](https://geo.music.apple.com/us/album/_/1477977279?i=1477977280&mt=1&app=music&at=1000lHKX)

And if I want to open the page on Odesly to check it (links could be changed) or submit feedback to developers for adding links — I can't do it easily.

I suggest to add a link to Odesly, probably even at the beginning, and it'll look like:

> Animadrop, Ben Walter, Aeris - Thin Air
> [Odesly](https://song.link/ru/i/1477977280) | [Deezer](https://www.deezer.com/track/738571332) | [Spotify](https://open.spotify.com/track/4HhvD1HVtvZ0eaCP0Sp0F0) | [Apple Music](https://geo.music.apple.com/us/album/_/1477977279?i=1477977280&mt=1&app=music&at=1000lHKX)

**Comment by 9dogs** [comment]
Yes, Odesli link would be helpful, though I'd rather not to put it on the first place since it'll generate less reasonable preview.

**Comment by AlexWayfer** [comment]
> though I'd rather not to put it on the first place since it'll generate less reasonable preview

Hm… are you sure? I see Deezer is the first now almost all the time, sometimes Google.

So, Deezer:

![image](https://user-images.githubusercontent.com/6510020/74725633-38610d80-524f-11ea-8cfe-9022d552155e.png)

Google:

![image](https://user-images.githubusercontent.com/6510020/74725678-49118380-524f-11ea-84b5-5a13d0cbc2b9.png)

Odesli:

![image](https://user-images.githubusercontent.com/6510020/74725781-6d6d6000-524f-11ea-93c4-f51fef789d86.png)

So, yeah, it's smaller… but:

1. It's more consistent, because we don't have links to Deezer or Google every time.
2. We can try to ask Odesli devs to change previews. I'll try to contact with them.

**Comment by AlexWayfer** [comment]
I've researched a difference and I found `og:type` with `website` value on Odesli and with `music.song` value on other music platforms. So, I've wrote about this to Odesli. I hope they'll change it.

**Comment by AlexWayfer** [comment]
I've got the answer from Odelsi staff:

> Hello Alex!
>
> I've looked into improving/changing our meta tags, and it's been on the backlog ever since. Technically, we don't host any music ourselves, so I don't think we should use the music.song type. I think we're more of a website, unfortunately.
>
> Best,
> Kurt

So… further as you wish. I want to see Odesli as the first link, but I understand your point, event with inconsistency of the first provider (Deezer, Google, etc.)

**Comment by kweiberth** [comment]
Thanks @AlexWayfer for bringing this to my attention. I'm hesitant to use the music.song and music.album tags bc we don't represent the music source itself. As you mentioned to me in our private chat, that might not matter so much bc technically we "represent" the song/album. I have two major concerns with making this change:

1. Changing the tag will change the previews, which may not be desired by others who use our service; i.e. other people might not like the big previews.
2. Since we don't actually host the music entities, they may not be semantically correct tags, and maybe we somehow get "punished" by Google/SEO?

I do think there's room for improvement. The current description we have is long and wordy. It's good as "advertisement" for our service, but it's not desirable for users like yourself, or probably even for recording artists. An alternative might be to get rid of everything past the emojis, so delete the "Powered by Odesli, ..." part.

**Comment by AlexWayfer** [comment]
@kweiberth thank you for coming and participation in public discussion!

> Changing the tag will change the previews, which may not be desired by others who use our service; i.e. other people might not like the big previews.

Debatable point. It's true, but we don't know. And if almost all music services have such preview — why Odesli can't? Also Odesli can collect such feedback after change, but I think this will be accepted by users as improvement. But it's just my opinion.

> Since we don't actually host the music entities, they may not be semantically correct tags, and maybe we somehow get "punished" by Google/SEO?

I'm pretty sure that no. 

1. Some non-public services, sometimes region-unavailable like Spotify, has no available song play. But they're indexed and showing in search results.
2. How search engine can check ability to play the audio? Current systems (platforms) even don't load files with clear media MIME-type while playing.
3. I've made a small research and found answers "no", it just affects visual display in social networks and such OG tags like `description` can be parsed by search engines. But we're talking primarily about `type` tag.
  * https://www.oncrawl.com/oncrawl-seo-thoughts/what-you-need-to-know-about-open-graph/
  * https://www.quora.com/Do-OpenGraph-meta-tags-have-an-impact-on-SEO
4. Does Odesli has SEO for search engines? I've searched directly for link from example (issue description) and didn't find anything at all:
![image](https://user-images.githubusercontent.com/6510020/74850307-dede0900-534a-11ea-8a78-023f16a7c989.png)
But maybe you talked about punishments for main page, by domain. But I still think "no".

**Comment by 9dogs** [comment]
As for now, links order is highly opinionated: Deezer is on the first place because I mainly use it as my service of choice (and Google Music as a sidekick). I'm deliberating if it'd be better to put free services first, mainly SoundCloud and YouTube.
Unfortunately, I don't see any way to find out the most frequent use case, for example how valuable would it be for a user to be able to listen track inside Telegram by clicking a preview.
Straightforward fix would be to make the order configurable but it will require a storage layer, so I'd like to consider other options first.

**Comment by AlexWayfer** [comment]
> As for now, links order is highly opinionated: Deezer is on the first place because I mainly use it as my service of choice (and Google Music as a sidekick). I'm deliberating if it'd be better to put free services first, mainly SoundCloud and YouTube.

Or they should be just alphabetical-sorted. But I'm still thinking that Odesli is the main, parent link.

> Straightforward fix would be to make the order configurable but it will require a storage layer, so I'd like to explore other options first.

Yes, like I've suggested in #2.

---

OK, about Odesli link position: it can be not the first, but on the separate line. If you don't like its preview — it can be no on the first line, but after the main, like:

> Animadrop, Ben Walter, Aeris - Thin Air
> [Deezer](https://www.deezer.com/track/738571332) | [Spotify](https://open.spotify.com/track/4HhvD1HVtvZ0eaCP0Sp0F0) | [Apple Music](https://geo.music.apple.com/us/album/_/1477977279?i=1477977280&mt=1&app=music&at=1000lHKX)
> [Odesli](https://song.link/ru/i/1477977280)

Maybe with empty lines between:

> Animadrop, Ben Walter, Aeris - Thin Air
>
> [Deezer](https://www.deezer.com/track/738571332) | [Spotify](https://open.spotify.com/track/4HhvD1HVtvZ0eaCP0Sp0F0) | [Apple Music](https://geo.music.apple.com/us/album/_/1477977279?i=1477977280&mt=1&app=music&at=1000lHKX)
>
> [Odesli](https://song.link/ru/i/1477977280)

---

Then, I suggest to order platforms links alphabetically, but as you wish. And it can be a separate issue. This issue is about Odesli link existence.

---

Then, @kweiberth can change OpenGraph tags for Odesli link. And then we can make it first. But it's not necessary, I guess — as Odesli staff wish.

**Comment by kweiberth** [comment]
I would suggest making the link to Odesli named "Listen on other platforms" or something similar, since users may not be familiar with Odesli. And I would say it probably should not be placed first bc it's a better UX if the user can click directly on the link to the platform as opposed to the extra click and page load going through Odesli. 

Thanks for researching the og tags. It does seem like we could use the music.* tags if we want to. Re: the SEO stuff, I'm mostly concerned about being blacklisted in some way by Google for our home page ranking. It's clearly not something that will obviously happen, but Google and SEO is also a black box and it's better to be cautious. 

At the end of the day, the og tags and link previews are a matter of taste and opinion. What are the advantages in your mind with the music.* previews? To me, the big difference seems to be that the album art is much larger. Personally, I would not prefer this type of preview in my messaging channels bc of how much room it takes up and how much thread context you lose.

As I mentioned, there are lots of people using our links, so this would be a big change. Unfortunately, I have no way right now of collecting feedback to try and figure out what would be preferred.

**Comment by AlexWayfer** [comment]
> I would suggest making the link to Odesli named "Listen on other platforms" or something similar, since users may not be familiar with Odesli.

1. It's literally "Odesli" bot.
2. On what "other" if bot includes all the links from Odesli API?

> Re: the SEO stuff, I'm mostly concerned about being blacklisted in some way by Google for our home page ranking. It's clearly not something that will obviously happen, but Google and SEO is also a black box and it's better to be cautious.

Acceptable, but unreasonable so far. 

> What are the advantages in your mind with the music.* previews? 

Large image (album cover) below title (and description?) with play button, you can see screenshots above. Maybe something else, in other platform, I'm not sure.

> Personally, I would not prefer this type of preview in my messaging channels bc of how much room it takes up and how much thread context you lose.

OK, if it's your personal taste.

> Unfortunately, I have no way right now of collecting feedback to try and figure out what would be preferred.

OK, thank you.

**Comment by kweiberth** [comment]
> On what "other" if bot includes all the links from Odesli API?

If you're rendering all of the links, then I don't see a necessity for the Odesli link. 

> Large image (album cover) below title (and description?) with play button, you can see screenshots above. Maybe something else, in other platform, I'm not sure.

We wouldn't be able to add a play button, bc we don't host the music ourselves, as I've mentioned. Seems like this boils down to a (personal) preference between small and large thumbnail. I also wonder how other platforms render previews, i.e. does Telegram render previews differently than Discord, Slack, Facebook, Twitter, etc. 

---

I appreciate you reaching out with this. I can tell you are disappointed, and I apologize for not being able to accommodate this request at this time.
